app/src/main/java/org/onionshare/android/files/FileManager.kt

```diff
@@ -0,0 +1,72 @@
+package org.onionshare.android.files
+
+import android.app.Application
+import android.content.Context.MODE_PRIVATE
+import android.net.Uri
+import android.text.format.Formatter.formatShortFileSize
+import android.util.Base64.NO_PADDING
+import android.util.Base64.URL_SAFE
+import android.util.Base64.encodeToString
+import androidx.documentfile.provider.DocumentFile
+import org.onionshare.android.R
+import org.onionshare.android.files.FileManager.State.FilesAdded
+import org.onionshare.android.files.FileManager.State.FilesReadyForDownload
+import org.onionshare.android.server.SendFile
+import java.io.File
+import java.util.zip.ZipEntry
+import java.util.zip.ZipOutputStream
+import javax.inject.Inject
+import kotlin.random.Random
+
+class FileManager @Inject constructor(
+    app: Application,
+) {
+
+    sealed class State {
+        object NoFiles : State()
+        open class FilesAdded(val files: List<SendFile>) : State()
+        class FilesReadyForDownload(files: List<SendFile>, val zip: File) : FilesAdded(files)
+    }
+
+    private val ctx = app.applicationContext
+
+    fun addFiles(uris: List<Uri>, state: State): FilesAdded {
+        val existingFiles = if (state is FilesAdded) state.files else emptyList()
+        val files = uris.mapNotNull { uri ->
+            // continue if we already have that file
+            if (existingFiles.any { it.uri == uri }) return@mapNotNull null
+
+            val documentFile = DocumentFile.fromSingleUri(ctx, uri) ?: error("Only API < 19")
+            val name = documentFile.name ?: uri.getFallBackName() ?: error("Uri has no path $uri")
+            val size = documentFile.length()
+            val sizeHuman = if (size == 0L) ctx.getString(R.string.unknown) else {
+                formatShortFileSize(ctx, size)
+            }
+            SendFile(name, sizeHuman, uri)
+        }
+        return FilesAdded(existingFiles + files)
+    }
+
+    fun zipFiles(state: FilesAdded): FilesReadyForDownload {
+        val zipFileName = encodeToString(Random.nextBytes(32), NO_PADDING or URL_SAFE).trimEnd()
```

Comment:
Nitpick: slightly different technique here from generating the static URL - maybe we want a utility function?

Reply:
I had used SecureRandom for the static path path and normal random here. Not sure if we need a secure random for the static path.If not, then yeah this could be a Utils function.